esp-link/cgitelnet.c

```diff
@@ -0,0 +1,85 @@
+#include <esp8266.h>
+#include "cgi.h"
+#include "config.h"
+#include "serbridge.h"
+
+// Cgi to return choice of Telnet ports
+int ICACHE_FLASH_ATTR cgiTelnetGet(HttpdConnData *connData) {
+  if (connData->conn==NULL) return HTTPD_CGI_DONE; // Connection aborted
+
+  char buff[1024];
+  int len;
+  
+  os_printf("Current telnet ports: port1=%d port2=%d\n",
+	flashConfig.telnet_port1, flashConfig.telnet_port2);
+	
+  len = os_sprintf(buff,
+      "{ \"port1\": \"%d\", \"port2\": \"%d\" }",
+      flashConfig.telnet_port1, flashConfig.telnet_port2);
+
+  jsonHeader(connData, 200);
+  httpdSend(connData, buff, len);
+  return HTTPD_CGI_DONE;
+}
+
+// Cgi to change choice of Telnet ports
+int ICACHE_FLASH_ATTR cgiTelnetSet(HttpdConnData *connData) {
+  if (connData->conn==NULL) {
+    return HTTPD_CGI_DONE; // Connection aborted
+  }
+	
+  char *coll;
+  int8_t ok = 0;
+  uint16_t port1, port2;
+  ok |= getUInt16Arg(connData, "port1", &port1);
+  ok |= getUInt16Arg(connData, "port2", &port2);
+  if (ok < 0) { coll = "Failed to fetch ports. Are they valid?"; goto collision; }
```

Comment:
This should perhaps be `<= 0`? As it is you're not returning any response when ok==0 and that's not good.
Second issue is that goto collision results in an rather poor error message.
Third issue is that having the string verbatim here ends up putting it into RAM, which is not plentiful. If you used sprintf the constant format string ends up in flash and doesn't consume RAM (there's some hacky magic that does that).